bf16-bitfury16.c

```diff
@@ -0,0 +1,1190 @@
+	uint32_t ret = 0x00000000;
+	uint8_t i;
+
+	for (i = 0; i < nbits; i++)
+		ret |= (((data >> i) & 0x1) << (nbits - (i + 1)));
+
+	return ret;
+}
+#endif
+
+uint32_t gen_mask(uint32_t nonce, uint8_t nbits) {
+	uint32_t mask = 0x00000000;
+
+	uint32_t mask_code = (nbits << 16);
+
+	/* highest 16 bits of nonce counter */
+	uint32_t nonce_code = (nonce & 0x0007ffff);
+#ifdef FLIP_BITS
+	uint32_t nonce_cntr = flip_bits(nonce_code, 19);
+	nonce_code = (flip_bits(nonce_cntr - 2, 19) ^ 0xaaaaaaaa);
+	mask = (mask_code | (nonce_code & (uint32_t)(pow(2, nbits) - 1)));
+#else
+	nonce_code = ((nonce_code ^ 0xaaaaaaaa) & (uint32_t)(pow(2, nbits) - 1));
+	mask = (mask_code | nonce_code);
+#endif
+
+	return mask;
+}
+
+void ms3steps16(uint32_t* p, uint32_t* w, uint32_t* task) {
+	uint32_t a, b, c, d, e, f, g, h, new_e, new_a;
+	uint8_t i;
+
+	a = p[0];
+	b = p[1];
+	c = p[2];
+	d = p[3];
+	e = p[4];
+	f = p[5];
+	g = p[6];
+	h = p[7];
+	for (i = 0; i < 3; i++) {
+		new_e = w[i] + sha256_k[i] + h + CH(e,f,g) + SHA256_F2(e) + d;
+		new_a = w[i] + sha256_k[i] + h + CH(e,f,g) + SHA256_F2(e) +
+		        SHA256_F1(a) + MAJ(a,b,c);
+		d = c;
+		c = b;
+		b = a;
+		a = new_a;
+		h = g;
+		g = f;
+		f = e;
+		e = new_e;
+	}
+
+	task[18] = ntohl(a ^ 0xaaaaaaaa);
+	task[17] = ntohl(b ^ 0xaaaaaaaa);
+	task[16] = ntohl(c ^ 0xaaaaaaaa);
+	task[15] = ntohl(d ^ 0xaaaaaaaa);
+	task[11] = ntohl(e ^ 0xaaaaaaaa);
+	task[10] = ntohl(f ^ 0xaaaaaaaa);
+	task[9]  = ntohl(g ^ 0xaaaaaaaa);
+	task[8]  = ntohl(h ^ 0xaaaaaaaa);
+}
+
+uint8_t gen_task_data(uint32_t* midstate, uint32_t merkle, uint32_t ntime, uint32_t nbits, uint32_t mask, uint8_t* task) {
+	uint8_t  i;
+	uint32_t tmp;
+	uint32_t w[3];
+
+	w[0] = merkle;
+	w[1] = ntime;
+	w[2] = nbits;
+
+	for (i = 0; i < 8; i++) {
+		tmp = midstate[i];
+		tmp ^= 0xaaaaaaaa;
+		tmp = ntohl(tmp);
+		cg_memcpy(task + i*4, &tmp, sizeof(tmp));
+	}
+
+	ms3steps16(midstate, w, (uint32_t*)task);
+
+	for (i = 0; i < 3; i++) {
+		tmp = w[i];
+		tmp ^= 0xaaaaaaaa;
+		tmp = ntohl(tmp);
+		cg_memcpy(task + (12 + i)*4, &tmp, sizeof(tmp));
+	}
+
+	mask = ntohl(mask);
+	cg_memcpy(task + 19*4, &mask, sizeof(mask));
+
+	return 0;
+}
+
+uint8_t spi_command_init(bf_command_t* command, const uint8_t depth, const bf_chip_address_t chip_address,
+		const bf_cmd_code_t cmd_code, const uint8_t data_length, const uint8_t* tx) {
+	uint8_t i;
+
+	memset(command->tx, 0, sizeof(command->tx));
+	memset(command->rx, 0, sizeof(command->rx));
+
+	command->tx[0] = 0x01;
+	command->data_length = 1;
+
+	if ((chip_address.chip_id > 10) && (chip_address.chip_id != 0x0f))
+		return 1;
+	else {
+		cg_memcpy(&command->chip_address, &chip_address, sizeof(bf_chip_address_t));
+		command->tx[1] = (command->chip_address.chip_id << 4);
+		command->data_length++;
+	}
+
+	command->depth = depth;
+
+	command->cmd_code = cmd_code;
+	command->tx[2] = command->cmd_code;
+	command->data_length++;
+
+	if ((data_length >=0) && (data_length <= 79)) {
```

Comment:
bf16-bitfury16.c:729:19: warning: comparison is always true due to limited range of data type [-Wtype-limits]
  if ((data_length >=0) && (data_length <= 79)) {